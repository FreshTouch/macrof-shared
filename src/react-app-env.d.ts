// eslint-disable-next-line spaced-comment
/// <reference types="@types/jest" />;
/// <reference types="react-scripts" />
/// <reference types="@emotion/react/types/css-prop" />

declare module '*.module.less' {
  const classes: { [key: string]: string };
  export default classes;
}

declare module '*.json' {
  const value: any;
  export default value;
}

declare module '*.css';
declare module '*.less';
