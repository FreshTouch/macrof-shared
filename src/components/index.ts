export * from './routers';
export * from './router';
