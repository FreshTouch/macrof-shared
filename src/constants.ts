import {_env } from "./index";

export const IS_DEV = _env('NODE_ENV') === 'development';

export const ERR = {
    CONTEXT: "[APP MF]: Context error!",
    BEFORE_INIT: "[MF Stores]: There was an attention of Stores usage before initiation!"
};
