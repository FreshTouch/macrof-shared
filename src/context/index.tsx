import {Container} from "inversify";
import React, {Context as ReactContext, useContext} from 'react';

import {
    CallBackProvider,
    TAppContextCallBack,
    TAppContextElement,
    TAppContextProps,
    TUseMFContext,
    TEmployeeStores,
    TEmployeeContext,
    createAppStores,
    IocIEmployeeStore,
    IRootStore,
    TMfStoreCb,
    TStores,
    TAppStoreContext,
    ERR
} from "..";

export function Context <TContext>(store?: TContext): ReactContext<TContext | null> {
    return React.createContext<TContext| null>(store || null)
}

export const AppStore: () => TEmployeeStores =
    ((store: TEmployeeStores): () => TEmployeeStores => (): TEmployeeStores => store)(createAppStores());

export const _appStoreContext: () => TEmployeeContext =
    ((context:TEmployeeContext): () => TEmployeeContext =>
      (): TEmployeeContext => context)(Context<TEmployeeStores>(AppStore()));

export const AppProvider: TAppContextCallBack<TAppContextElement> =
    ((AppContext: TEmployeeContext, appStore: () => TEmployeeStores): TAppContextCallBack<JSX.Element> =>
        ({...props}: TAppContextProps): JSX.Element =>
          <AppContext.Provider value={ appStore() }>{props.children}</AppContext.Provider>
)(_appStoreContext(), AppStore);

export function useAppStoreContext(): TEmployeeStores {
    const stores = React.useContext<TEmployeeStores | null>(_appStoreContext());

    if (!stores) {
        throw new Error(ERR.BEFORE_INIT);
    }

    return stores;
}

export function useMfContext(): TUseMFContext {
    const context: TEmployeeStores = useAppStoreContext();
    return { ...context.employeeStore.init(), ...context.rootStore.init() } as TUseMFContext;
}

export const useEmpStoreContext:() => IocIEmployeeStore =
    (): IocIEmployeeStore => useAppStoreContext().employeeStore;

export const useRootStoreContext: () => IRootStore =
    (): IRootStore => useAppStoreContext().rootStore

export const initMFProviders: (provider: CallBackProvider) => void =
    (provider: CallBackProvider): void => AppStore().employeeStore.inject(provider);

export function configureContextStore<T>(mfStoreCb: TMfStoreCb<T>): TStores<T> {
    const generalContextStore: TAppStoreContext =
        AppStore().rootStore as TAppStoreContext;

    if (!generalContextStore?.rootContainer) {
        console.error(ERR.CONTEXT);
    }

    const container: Container = new Container({
        defaultScope: 'Singleton',
        autoBindInjectable: true
    });

    container.parent = generalContextStore.rootContainer;

    return {
        ...generalContextStore.init(),
        ...mfStoreCb(container)
    };
}

export function useConfiguredStores<T>(configuredContext: ReactContext<TStores<T> | null>): TStores<T> {
    const context = useContext(configuredContext);

    if (!context) {
        console.error(ERR.BEFORE_INIT);
    }

    return context as TStores<T>;
}
