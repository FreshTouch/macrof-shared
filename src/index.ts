import "reflect-metadata";

export * from "./context/mfHistory";

export * from "./utils";
export * from "./constants";
export * from "./models";
export * from "./ioc";
export * from "./services";
export * from "./context";
export * from "./components";
