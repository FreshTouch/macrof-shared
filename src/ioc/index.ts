import {MF_TYPES} from "./types";

export * from "./createStores";
export * from "./interfaces";
export * from "./initial";

export {
  MF_TYPES
}
