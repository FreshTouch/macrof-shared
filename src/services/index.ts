export * from './logger';

export * from './http/response';
export * from './http/interface';
export * from './http/middleware';
export * from './http/service';

export * from './loader';
