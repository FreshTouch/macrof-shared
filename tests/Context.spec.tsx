import { render, renderHook } from "@testing-library/react";
import React from "react";

import {AppProvider, AppStore, useMfContext} from '../src';

const AppContext = (): JSX.Element => (
  <AppProvider>
    <div>Test</div>
  </AppProvider>
);

describe("Context", () => {
  describe("Provider", () => {
    it("AppProvider should render successfully", () => {
      const { baseElement } = render(<AppContext />);
      expect(baseElement).toMatchSnapshot();
      expect(baseElement).toBeTruthy();
    });
  });

  describe("AppStore", () => {
    it("Stores must be initialized correctly", () => {
      const { current } = renderHook(useMfContext).result;
      expect(current).toMatchObject(AppStore().rootStore.init());
    });
  });
});
