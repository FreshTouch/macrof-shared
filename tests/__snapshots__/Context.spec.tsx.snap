// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Provider AppProvider should render successfully 1`] = `
<body>
  <div>
    <div>
      Test
    </div>
  </div>
</body>
`;
